### PR TITLE
 [FEATURE] [OGR provider] Load/save style in database for GPKG and Spatialite

### PR DESCRIPTION
**Pull request by rouault**

### CODE REVIEW

File: python/core/qgsvectorlayer.sip [hunk 1]
```diff
@@ -662,11 +662,11 @@ class QgsVectorLayer : QgsMapLayer
      * @param description
      * @param useAsDefault
      * @param uiFileContent
-     * @param msgError
+     * @param msgError (out)
      */
     virtual void saveStyleToDatabase( const QString& name, const QString& description,
                                       bool useAsDefault, const QString& uiFileContent,
-                                      QString &msgError );
+                                      QString &msgError /Out/ );
```
Comment (m-kuhn) [inline python/core/qgsvectorlayer.sip]: Should that go to api_break.dox?

Comment (rouault) [inline python/core/qgsvectorlayer.sip]: I don't think so. The API was broken before. It's an API fix :-)

File: src/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -3545,6 +3576,497 @@ bool QgsOgrProvider::leaveUpdateMode()
...
+                                 layerName,
+                                 subsetString,
+                                 ogrGeometryType );
+
+  OGRDataSourceH hDS = QgsOgrProviderUtils::OGROpenWrapper( TO8F( filePath ), true, nullptr );
+  if ( !hDS )
+  {
+    QgsDebugMsg( "Connection to database failed.." );
+    errCause = QObject::tr( "Connection to database failed" );
+    return NULL;
```
Comment (m-kuhn) [inline src/providers/ogr/qgsogrprovider.cpp]: Use `nullptr`

File: python/core/qgsvectorlayer.sip (hunk 1 above)
Comment (m-kuhn) [inline python/core/qgsvectorlayer.sip]: Nice way to put it, `mv api_break.dox api_fix.dox` 😝

Comment (rouault) [inline python/core/qgsvectorlayer.sip]: Well I was serious. The API was non functional before (or at the very least seriously cumbersome). I doubt anyone used that method before.

Comment (m-kuhn) [inline python/core/qgsvectorlayer.sip]: I know that it was non-functional. Fact is, *if* someone used it, for him it's conceptually an API break. And API breaks should be documented as good as possible. I guess quite some functions in this document are probably unused. But on the other hand, I'm sure we would all wonder what actually gets used somewhere out there.

Comment (rouault) [inline python/core/qgsvectorlayer.sip]: OK. I just documented the changes

**Comment by rduivenvoorde** [comment]
Whee... yes yes!

**Comment by rduivenvoorde** [comment]
@rouault and @pka
FYI Sourcepole also showed a nice (experimental?) plugin at FOSS4G which not only saves the styles, but a full project including images in a GPKG: 

http://plugins.qgis.org/plugins/QgisGeopackage/
https://github.com/pka/qgpkg

I do not think this is related, but just wanted to let you both know of each others work :-)

**Comment by m-kuhn** [comment]
Thanks

**Pull request merged**